src/node_sea.cc

```diff
@@ -238,6 +318,49 @@ std::optional<SeaConfig> ParseSingleExecutableConfig(
   return result;
 }
 
+std::optional<std::string> GenerateCodeCache(std::string_view main_path,
+                                             std::string_view main_script) {
+  RAIIIsolate raii_isolate;
+  Isolate* isolate = raii_isolate.get();
+
+  HandleScope handle_scope(isolate);
+  Local<Context> context = Context::New(isolate);
+  Context::Scope context_scope(context);
+
+  errors::PrinterTryCatch bootstrapCatch(
+      isolate, errors::PrinterTryCatch::kPrintSourceLine);
+
+  Local<String> filename;
+  if (!String::NewFromUtf8(isolate, main_path.data()).ToLocal(&filename)) {
+    return std::nullopt;
+  }
+
+  Local<String> content;
+  if (!String::NewFromUtf8(isolate, main_script.data()).ToLocal(&content)) {
+    return std::nullopt;
+  }
+
+  std::vector<Local<String>> parameters = {
+      FIXED_ONE_BYTE_STRING(isolate, "exports"),
+      FIXED_ONE_BYTE_STRING(isolate, "require"),
+      FIXED_ONE_BYTE_STRING(isolate, "module"),
+      FIXED_ONE_BYTE_STRING(isolate, "__filename"),
+      FIXED_ONE_BYTE_STRING(isolate, "__dirname"),
+  };
+
+  Local<Function> fn;
+  if (!contextify::CompileFunction(
```

Comment:
Note that generating the code cache here directly is not completely safe until we have https://chromium-review.googlesource.com/c/v8/v8/+/4543306 in main. But since this is just an experimental feature it's fine as long as our tests don't crash for the time being. It might also not be completely safe until V8 stops serializing host defined options with the scripts, but I don't think we really support `import()` in SEA for now so that's probably fine?

Reply:
I tried using `import()` in an SEA with the code cache changes and it throws a `TypeError: Invalid host defined options` error from here - https://github.com/nodejs/node/blob/2fca7ea2be160d8c30c5bacf7d6246e4c6f2a7f0/src/module_wrap.cc#L576 but it worked with no errors when the code cache wasn't being used, i.e., when using the Node.js binary from `main`.

Reply:
@joyeecheung I've added a TODO to add support for `import()`, PTAL.

Reply:
Agh, I wasn't aware that `import()` didn't work with code caching when I suggested this, sorry :\

Do we know if there's other things we might want to support that don't work with code caching? Worst-case: we can add another configuration knob to the SEA config to allow the user to choose whether or not they want code caching.

Reply:
Ah sorry I didn't notice that it is set to true by default here.

> Not really sure what should be done.

I think for now we could just leave this false by default. Until the known issues are addressed this could still be sort of "use it at your own risk" feature.

> Do we know if there's other things we might want to support that don't work with code caching? 

The two above are what I know off the top of my head. But in general if they don't work V8 usually just fails silently.

Reply:
I've turned this into an opt-in feature. PTAL @joyeecheung @kvakil